### PR TITLE
Upgrade mlx-lm and update to use new BatchedGeneration API

### PR DESCRIPTION
**Pull request by neilmehta24**
- Add an adapter for the logits processors for batched generation. This is needed right now due to a regression in mlx-lm in the logics processor API. This adapter can be removed after this PR: https://github.com/ml-explore/mlx-lm/pull/1115
- Start using the `all_tokens` API so that the logits processors see the entire history
- Switch to the new `batch_generator.next()` API, and get rid of the old `progress_callback` definition
- Switch to the new `stop_tokens` API
- Add helper function that trims cache by one to create a seed token on an exact cache hit.

Codex review:
Screenshot 2026-04-06 at 4 06 17 PM

### CODE REVIEW

File: mlx_engine/model_kit/batched_model_kit.py [hunk 1]
```diff
@@ -27,6 +28,81 @@
...
+    if len(prompt_tokens) == 0:
+        return cache, [], []
+
+    # Exact cache hits need one token moved out of cache to start decoding.
+    if cache is not None and can_trim_prompt_cache(cache):
+        trim_prompt_cache(cache, 1)
+        return cache, prompt_tokens[:-1], prompt_tokens[-1:]
+
+    # If we cannot trim an exact-hit cache, replay the full prompt instead.
+    return None, [], prompt_tokens
```
Comment (will-lms) [inline mlx_engine/model_kit/batched_model_kit.py:103]: Can we do better? Could we search for `prompt_tokens[:-1]` instead of the full length and then need to trim one off?

Comment (neilmehta24) [reply mlx_engine/model_kit/batched_model_kit.py:103]: The `fetch_nearest_cache` API can behave surprisingly. There are three cases to consider. Let's assume `[1,2,3,4]` is cached, and our prompt is `[1,2,3,4]`

- Query `[1,2,3]`. The API returns cache=`[1,2]` rest=`[3]`. We would then also append `4` to `rest`
- Query `[1,2,3,4]`. The API returns cache=`[1,2,3,4]` rest=`[]`
- Query `[1,2,3,4,5]`. The API returns cache=`[1,2,3,4]` rest=`[5]`

If we were to query `prompt_tokens[-1]`, we would use case 1 and have an extra token trimmed off.

I imagine this API was designed to avoid trimming the cache unless necessary, which is why it takes on this behavior and why we need to trim on exact matches.

Planning on keeping this as-is.

File: mlx_engine/model_kit/batched_model_kit.py [hunk 2]
```diff
@@ -27,6 +28,81 @@
...
+        return cache, prompt_tokens[:cached_prefix_len], rest
+
+    # Empty prompts have no seed token to preserve.
+    if len(prompt_tokens) == 0:
+        return cache, [], []
+
+    # Exact cache hits need one token moved out of cache to start decoding.
+    if cache is not None and can_trim_prompt_cache(cache):
+        trim_prompt_cache(cache, 1)
+        return cache, prompt_tokens[:-1], prompt_tokens[-1:]
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/model_kit/batched_model_kit.py]: fair enough, added check.

**Review by will-lms — APPROVED** [review]
